rse-py-package.Rmd

```diff
@@ -0,0 +1,1460 @@
+5. Exit the python interpreter, and deactivate the `test` environment in the other terminal.
+
+## How can I create an installable Python package? {#rse-py-package-package}
+
+People can always get your package by cloning your repository and copying files from that
+(assuming your repository is accessible,
+which is should be for published research),
+but it's much friendlier to create something they can install
+(and it will be easier for us when it comes to running test suites
+and distributing, too!)
+
+For historical reasons,
+Python has several ways to build an installable package.
+We will show how to use [setuptools][setuptools],
+which is the lowest common denominator;
+[conda][conda] is a modern does-everything solution,
+but has larger startup overhead.
+
+To use `setuptools`,
+we must create a file called `setup.py` in the directory *above* the root directory of the package:
+
+```text
++- setup.py
++- test_zipfpy.py
++- zipfpy
+    +- __init__.py
+    +- check.py
+    +- generate.py
+```
+
+<!-- == noindent -->
+The file `setup.py` must have exactly that name,
+and must contain these lines:
+
+<!-- src="rse-py-package/06/setup.py" -->
+```python
+from setuptools import setup
+
+setup(
+    name='zipfpy',
+    version='0.1',
+    author='Greg Wilson',
+    packages=['zipfpy']
+)
+```
+
+<!-- used="rse-py-package/06/test_zipfpy.py" -->
+<!-- used="rse-py-package/06/zipfpy/__init__.py" -->
+<!-- used="rse-py-package/06/zipfpy/check.py" -->
+<!-- used="rse-py-package/06/zipfpy/generate.py" -->
+
+This is enough to be able to install the package from the
+local copy source code using standard tools!
+
+Let's create a virtual environment to test the installation:
+
+```text
+$ virtualenv test_zipfpy
+...
+$ source test_zipfpy/bin/activate
+(test_zipfpy) $ pip install .
+Processing /Users/pterry/code
+Building wheels for collected packages: zipfpy
+  Building wheel for zipfpy (setup.py) ... done
+  Stored in directory: /private/var/folders/44/[etc]
+Successfully built zipfpy
+Installing collected packages: zipfpy
+Successfully installed zipfpy-0.1
+```
+
+Success!  If you take a look in `test_zipfpy/lib/python3.7/site-packages/`, you will
+now see the package with all the other site packages.
+
+We can generate Zipf-distributed data to our hearts content:
+
+```python
+>>> import zipfpy
+>>> import zipfpy.generate
+>>> zipf_out = zipfpy.generate.make_zipf(5)
+>>> print(zipf_out)
+[1.0, 0.5, 0.3333333333333333, 0.25, 0.2]
+```
+
+But if we continue and try to check the output, we'll find a problem:
+
+```python
+>>> import zipfpy.check
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+  File "/Users/pterry/code/zipfpy/check.py", line 2, in <module>
+    from pytest import approx
+ModuleNotFoundError: No module named 'pytest'
+```
+
+Ah ha - we're missing a dependency, we use `pytest` for approximate
```

Comment:
Feels like the sentence should end after the comma.

Reply:
Done